lib/clowne/utils/operation.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require 'clowne/utils/clone_mapper'
+
+module Clowne
+  module Utils
+    class Operation # :nodoc: all
+      THREAD_KEY = :"#{name}.clowne_operation"
+      DEFAULT_MAPPER = Utils::CloneMapper
+
+      private_constant :THREAD_KEY
+
+      class << self
+        def current
+          Thread.current[THREAD_KEY]
+        end
+
+        def wrap(operation_class: self, mapper: nil)
+          return yield if current
+
+          Thread.current[THREAD_KEY] = operation_class.new(mapper || DEFAULT_MAPPER.new)
+
+          current.tap do |operation|
+            operation.clone = yield
+            clear!
+          end
+        end
+
+        def clear!
+          Thread.current[THREAD_KEY] = nil
+        end
+      end
+
+      attr_accessor :clone
+      attr_reader :mapper
+
+      def initialize(mapper)
+        @post_processings = []
+        @mapper = mapper
+      end
+
+      def add_post_processing(block)
+        @post_processings.unshift(block)
+      end
+
+      def add_mapping(origin, clone)
+        @mapper.add(origin, clone)
+      end
+
+      def save
+        @clone.save
+      end
+
+      def save!
+        @clone.save!
+      end
+
+      def do_post_processing!
```

Comment:
Is there a reason of calling `do_post_porcessing!` as a separate step and not the part of the `save!` call?

As I see in the [example](https://github.com/onboardiq/fountain/compare/chore/stage-cloner...chore/stage-cloner-2), we always call `do_post_processing!` after `save!`.

I suggest something like this:

```ruby
def commit!
  # wrap into transaction
  clone_result.transaction do
    clone_result.save!
    do_post_processing
  end
end

def commit
  clone_result.transaction do
    do_post_processing if clone_result.save
  end
end
```

Reply:
It can simply be implemented only for ActiveRecord :( Sequel (ROM) has another way to use transaction so it will depend on an adapter and I not sure that it is a great idea. And also I cannot to say that _ALL_ users want to execute `post_processing` with a transaction. Yep, its general approach but not the only one and to provide the more flexible solution (when user wrap to transaction outside gem) sounds better for me. For example, a user can save clone and accumulate errors from `post_processing` blocks. Or uses `fail fast` strategy and aborts all next `post_processing` blocks but keeps origin clone in DB. (all examples will be possible when I'll implement parameters for post-processing blocks).
And the last: `do_post_processing!` it is **really** dangerous operation which can change a lot of records and I think we should provide explicit execution for it.
Anyway, what do you think about "sugar" via kwargs:

```ruby
def commit!(do_post_processing: false)
# or def commit!(after_commit: false)
end
```
?

Reply:
> 👍 I like commit(skip_after_clone: true)

 Hm, your key point is execution `after_clone` is the default behavior, am I right?